File: scripts/test-harden.ts
/**
 * T16 hardening (host / UI): FrameDecoder, IPC path, Trace/Graph caps,
 * supervisor restart under load.
 *
 * Synthetic path always runs. Live vcan restart-under-RX/TX is SKIP if
 * vcan0 is not UP.
 */
import assert from 'node:assert/strict'
import { spawnSync } from 'node:child_process'
import { chmodSync, mkdtempSync } from 'node:fs'
import { tmpdir } from 'node:os'
import { join } from 'node:path'
import { test } from 'node:test'
import { MAX_PENDING_REQUESTS } from '../electron/main/engineClient'
import { encodeMessage, FrameDecoder, MAX_FRAME_BYTES, ProtocolError } from '../electron/main/framing'
import { createIpcSocketPath } from '../electron/main/ipcPath'
import { EngineSupervisor } from '../electron/main/engineSupervisor'
import type { FrameEvent, RxBatch } from '../shared/engine'
import { GraphStore } from '../shared/graphStore'
import { GRAPH_CATALOG_MAX, GRAPH_SERIES_MAX, maxSamplesForWindow } from '../shared/graphWindow'
import { applyRxBatch } from '../shared/traceControl'
import { FrameRing, TRACE_RING_CAPACITY } from '../shared/traceRing'

function makeFrame(overrides: Partial<FrameEvent> = {}): FrameEvent {
  return {
    busId: 'bus-1',
    ifName: 'vcan0',
    ts_us: 1_700_000_000_000_000,
    can_id: 0x100,
    dlc: 8,
    data: 'e8035a0a00000000',
    is_eff: false,
    is_fd: false,
    brs: false,
    is_rtr: false,
    is_err: false,
    dir: 'rx',
    rate_ms: 10,
    decode: {
      name: 'EngineStatus',
      signals: { EngineSpeed: 250, EngineTemp: 50 },
      units: { EngineSpeed: 'rpm', EngineTemp: 'degC' }
    },
    ...overrides
  }
}

function waitFor(
  supervisor: EngineSupervisor,
  connected: boolean,
  timeoutMs: number
): Promise<void> {
  if (supervisor.getStatus().connected === connected) {
    return Promise.resolve()
  }
  return new Promise((resolve, reject) => {
    const timer = setTimeout(() => {
      unsubscribe()
      reject(new Error(`timed out waiting for connected=${connected}`))
    }, timeoutMs)
    const unsubscribe = supervisor.onStatus((status) => {
      if (status.connected === connected) {
        clearTimeout(timer)
        unsubscribe()
        resolve()
      }
    })
  })
}

function vcan0Up(): boolean {
  const flags = spawnSync('cat', ['/sys/class/net/vcan0/flags'], { encoding: 'utf8' })
  const type = spawnSync('cat', ['/sys/class/net/vcan0/type'], { encoding: 'utf8' })
  if (flags.status !== 0 || type.status !== 0) {
    return false
  }
  if (type.stdout.trim() !== '280') {
    return false
  }
  const value = Number.parseInt(flags.stdout.trim(), 16)
  return Number.isFinite(value) && (value & 1) === 1
}

function injectVcanFrame(canId: number, dataHex: string): void {
  const script = `
import sys
try:
    import can
except ImportError:
    sys.exit(2)
bus = can.Bus(interface="socketcan", channel="vcan0")
bus.send(can.Message(arbitration_id=${canId}, data=bytes.fromhex("${dataHex}"), is_extended_id=False))
bus.shutdown()
`
  const viaPython = spawnSync('python3', ['-c', script], { encoding: 'utf8' })
  if (viaPython.status === 0) {
    return
  }
  const viaCansend = spawnSync(
    'cansend',
    ['vcan0', `${canId.toString(16).toUpperCase().padStart(3, '0')}#${dataHex}`],
    { encoding: 'utf8' }
  )
  if (viaCansend.status !== 0) {
    throw new Error(viaPython.stderr || viaCansend.stderr || 'frame inject failed')
  }
}

test('FrameDecoder rejects oversized / zero / malformed and buffers partials', () => {
  const decoder = new FrameDecoder()
  const hello = encodeMessage({ type: 'engine.hello', payload: { name: 'x' } })
  assert.deepEqual(decoder.push(hello.subarray(0, 3)), [])
  const rest = decoder.push(hello.subarray(3))
  assert.equal(rest.length, 1)
  assert.equal(rest[0]?.type, 'engine.hello')

  const over = Buffer.alloc(4)
  over.writeUInt32BE(MAX_FRAME_BYTES + 1, 0)
  const overDec = new FrameDecoder()
  assert.throws(
    () => overDec.push(over),
    (error: unknown) => error instanceof ProtocolError && error.code === 'payload_too_large'
  )

  const zero = Buffer.alloc(4)
  const zeroDec = new FrameDecoder()
  assert.throws(
    () => zeroDec.push(zero),
    (error: unknown) => error instanceof ProtocolError && error.code === 'invalid_length'
  )

  const badJson = Buffer.from('{not-json', 'utf8')
  const header = Buffer.alloc(4)
  header.writeUInt32BE(badJson.length, 0)
  const badDec = new FrameDecoder()
  assert.throws(
    () => badDec.push(Buffer.concat([header, badJson])),
    (error: unknown) => error instanceof ProtocolError && error.code === 'invalid_json'
  )
})

test('createIpcSocketPath stays off world-writable XDG', () => {
  const dirty = mkdtempSync(join(tmpdir(), 'vb-world-'))
  chmodSync(dirty, 0o777)
  const prev = process.env.XDG_RUNTIME_DIR
  process.env.XDG_RUNTIME_DIR = dirty
  try {
    const path = createIpcSocketPath(process.pid)
    assert.ok(!path.startsWith(dirty + '/'), `must not use world-writable XDG: ${path}`)
    assert.ok(path.includes('vanillabus-') || path.includes('engine.sock'))
  } finally {
    if (prev === undefined) {
      delete process.env.XDG_RUNTIME_DIR
    } else {
      process.env.XDG_RUNTIME_DIR = prev
    }
  }
  assert.equal(MAX_PENDING_REQUESTS, 128)
})

test('Trace ring holds 20k under a 25k synthetic flood', () => {
  assert.equal(TRACE_RING_CAPACITY, 20_000)
  const ring = new FrameRing(TRACE_RING_CAPACITY)
  const flood = 25_000
  const frames: FrameEvent[] = []
  for (let index = 0; index < flood; index += 1) {
    frames.push(
      makeFrame({
        ts_us: 1_700_000_000_000_000 + index,
        can_id: 0x100 + (index % 64)
      })
    )
  }
  const dropped = applyRxBatch(ring, frames, false)
  assert.equal(ring.size, TRACE_RING_CAPACITY)
  assert.equal(dropped, flood - TRACE_RING_CAPACITY)
  assert.equal(ring.droppedOldest, flood - TRACE_RING_CAPACITY)
  assert.equal(ring.at(0).frame.ts_us, 1_700_000_000_000_000 + (flood - TRACE_RING_CAPACITY))
})

test('Graph sample window and series caps hold under flood', () => {
  const store = new GraphStore()
  store.setHz(20)
  store.setWindowSec(10)
  store.setSelected(['Flood0.Sig'])

  const start = 5_000_000_000_000
  const periodUs = 1_000
  const frames: FrameEvent[] = []
  for (let index = 0; index < 8_000; index += 1) {
    const series = index < 400 ? index : index % 40
    frames.push(
      makeFrame({
        ts_us: start + index * periodUs,
        can_id: series,
        decode: {
          name: `Flood${series}`,
          signals: { Sig: index },
          units: {}
        }
      })
    )
  }
  store.appendBatch({ frames, dropped: 17 })
  assert.equal(store.engineDropped, 17)
  assert.ok(store.seriesCount() <= GRAPH_SERIES_MAX, `series ${store.seriesCount()} > ${GRAPH_SERIES_MAX}`)
  assert.ok(
    store.catalogEntries().length <= GRAPH_CATALOG_MAX,
    `catalog ${store.catalogEntries().length} > ${GRAPH_CATALOG_MAX}`
  )
  assert.ok(
    store.maxSeriesLength() <= maxSamplesForWindow(10, 20),
    `samples ${store.maxSeriesLength()} over window bound`
  )
  assert.ok(store.sampleCount('Flood0.Sig') > 0)
})

test(
  'supervisor restart under IPC load reconnects without hung requests',
  { timeout: 25_000 },
  async (t) => {
    if (process.platform !== 'linux') {
      t.skip('engine supervisor is Linux-only')
      return
    }

    const supervisor = new EngineSupervisor()
    supervisor.start()
    try {
      await waitFor(supervisor, true, 10_000)
      const inflight: Promise<string>[] = []
      const load = setInterval(() => {
        inflight.push(supervisor.listBuses().then(() => 'ok', () => 'err'))
      }, 8)
      await new Promise((resolve) => setTimeout(resolve, 50))
      const pid = supervisor.getEnginePid()
      assert.ok(pid, 'expected a live engine PID')
      process.kill(pid, 'SIGKILL')
      await waitFor(supervisor, false, 5_000)
      await waitFor(supervisor, true, 10_000)
      clearInterval(load)
      const settled = await Promise.all(inflight)
      assert.ok(settled.every((item) => item === 'ok' || item === 'err'))
      const listed = await supervisor.listBuses()
      assert.ok(Array.isArray(listed.interfaces))
    } finally {
      supervisor.stop()
    }
  }
)

test(
  'supervisor restart under vcan RX/TX load resumes (SKIP if no vcan)',
  { timeout: 30_000 },
  async (t) => {
    if (process.platform !== 'linux') {
      t.skip('engine supervisor is Linux-only')
      return
    }
    if (!vcan0Up()) {
      t.diagnostic('SKIP vcan0 harden: vcan0 is not UP. sudo ./scripts/setup-vcan.sh')
      console.log('SKIP vcan0 harden: vcan0 is not UP on this host.')
      return
    }

    const supervisor = new EngineSupervisor()
    const batches: RxBatch[] = []
    const offRx = supervisor.onRxBatch((batch) => {
      batches.push(batch)
    })
    supervisor.start()
    try {
      await waitFor(supervisor, true, 10_000)
      const listed = await supervisor.listBuses()
      const vcan = listed.interfaces.find((iface) => iface.name === 'vcan0')
      if (!vcan || vcan.state !== 'up') {
        t.diagnostic('SKIP vcan0 harden: vcan0 is not UP. sudo ./scripts/setup-vcan.sh')
        console.log('SKIP vcan0 harden: vcan0 is not UP on this host.')
        return
      }

      const opened = await supervisor.openBus('vcan0')
      const cyclic = await supervisor.startCyclic({
        busId: opened.busId,
        can_id: 0x5A2,
        data: '01020304',
        period_ms: 10
      })
      for (let index = 0; index < 20; index += 1) {
        injectVcanFrame(0x42C, 'aabbccdd')
      }

      const pid = supervisor.getEnginePid()
      assert.ok(pid, 'expected a live engine PID')
      process.kill(pid, 'SIGKILL')
      await waitFor(supervisor, false, 5_000)
      await waitFor(supervisor, true, 10_000)

      batches.length = 0
      const reopened = await supervisor.openBus('vcan0')
      assert.ok(reopened.busId)
      injectVcanFrame(0x42D, '11223344')
      const deadline = Date.now() + 500
      let matched = false
      while (Date.now() < deadline && !matched) {
        matched = batches.some((batch) =>
          batch.frames.some((frame) => frame.can_id === 0x42d && frame.dir === 'rx')
        )
        if (!matched) {
          await new Promise((resolve) => setTimeout(resolve, 10))
        }
      }
      assert.equal(matched, true, 'expected rx.batch after restart-under-load reopen')
      assert.ok(typeof cyclic.job_id === 'string')
      await supervisor.closeBus(reopened.busId)
    } finally {
      offRx()
      supervisor.stop()
    }
  }
)

File: scripts/test-harden.py
#!/usr/bin/env python3
"""T16 hardening: RX backpressure, OOM caps, IPC safety, restart under load.

Synthetic / in-process path always runs (no host CAN required). If vcan0 is
UP a live flood + engine-kill slice also runs. Otherwise:

    SKIP vcan0 harden: vcan0 is not UP on this host.
    Bring it up with: sudo ./scripts/setup-vcan.sh
"""

from __future__ import annotations

import os
import queue
import socket
import stat
import struct
import subprocess
import sys
import tempfile
import threading
import time
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
ENGINE = ROOT / "engine"
sys.path.insert(0, str(ENGINE))

from can_engine.bus import BusManager  # noqa: E402
from can_engine.dbc import DbcError, resolve_dbc_path  # noqa: E402
from can_engine.framing import (  # noqa: E402
    MAX_FRAME_BYTES,
    ProtocolError,
    check_length,
    decode_payload,
    encode_message,
)
from can_engine.protocol import rx_batch_message  # noqa: E402
from can_engine.rate import RATE_KEY_MAX, RateTracker  # noqa: E402
from can_engine.rx import (  # noqa: E402
    RX_BATCH_INTERVAL_S,
    RX_BATCH_MAX_FRAMES,
    RX_BATCH_MAX_INTERVAL_S,
    RX_QUEUE_MAX,
    RxPump,
    should_flush,
)
from can_engine.server import IPC_PARTIAL_READ_TIMEOUT_S  # noqa: E402


class FakeMsg:
    def __init__(self, arbitration_id: int, data: bytes) -> None:
        self.arbitration_id = arbitration_id
        self.data = data
        self.is_extended_id = False
        self.is_fd = False
        self.bitrate_switch = False
        self.is_remote_frame = False
        self.is_error_frame = False
        self.dlc = len(data)


class ScriptedBus:
    def __init__(self) -> None:
        self._q: queue.Queue[FakeMsg] = queue.Queue()
        self.shutdowns = 0

    def inject(self, msg: FakeMsg) -> None:
        self._q.put(msg)

    def recv(self, timeout: float | None = None):
        try:
            return self._q.get(timeout=timeout if timeout is not None else 0.05)
        except queue.Empty:
            return None

    def shutdown(self) -> None:
        self.shutdowns += 1


def _fake_sysfs_vcan0(root: Path) -> Path:
    net = root / "net"
    virt = root / "devices" / "virtual" / "net" / "vcan0"
    virt.mkdir(parents=True)
    (virt / "type").write_text("280\n")
    (virt / "flags").write_text("0x41\n")
    net.mkdir()
    (net / "vcan0").symlink_to(virt)
    return net


def _choose_socket_path() -> Path:
    xdg = os.environ.get("XDG_RUNTIME_DIR")
    if xdg:
        parent = Path(xdg)
        mode = parent.stat().st_mode
        if not mode & stat.S_IWOTH:
            path = parent / f"vanillabus-harden-test-{os.getpid()}.sock"
            if path.exists():
                path.unlink()
            return path
    directory = Path(tempfile.mkdtemp(prefix="vanillabus-harden-"))
    os.chmod(directory, 0o700)
    return directory / "engine.sock"


def _recv_message(sock: socket.socket, timeout: float = 3.0) -> dict:
    sock.settimeout(timeout)
    header = b""
    while len(header) < 4:
        chunk = sock.recv(4 - len(header))
        if not chunk:
            raise RuntimeError("engine closed the socket while reading length")
        header += chunk
    length = int.from_bytes(header, "big")
    check_length(length)
    payload = b""
    while len(payload) < length:
        chunk = sock.recv(length - len(payload))
        if not chunk:
            raise RuntimeError("engine closed the socket while reading JSON")
        payload += chunk
    return decode_payload(payload)


def _recv_until(sock: socket.socket, types: set[str], timeout: float = 2.0) -> dict:
    deadline = time.time() + timeout
    while True:
        remaining = deadline - time.time()
        if remaining <= 0:
            raise TimeoutError(f"timed out waiting for {sorted(types)}")
        message = _recv_message(sock, timeout=remaining)
        if message.get("type") in types:
            return message


def _send_message(sock: socket.socket, message: dict) -> None:
    sock.sendall(encode_message(message))


def _start_engine(ipc_path: Path) -> subprocess.Popen[bytes]:
    env = os.environ.copy()
    existing = env.get("PYTHONPATH", "")
    env["PYTHONPATH"] = str(ENGINE) if not existing else f"{ENGINE}{os.pathsep}{existing}"
    proc = subprocess.Popen(
        [sys.executable, "-m", "can_engine", "--ipc", str(ipc_path)],
        cwd=str(ROOT),
        env=env,
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
    )
    deadline = time.time() + 5
    while time.time() < deadline:
        if proc.poll() is not None:
            output = proc.stdout.read().decode("utf-8", errors="replace") if proc.stdout else ""
            raise RuntimeError(f"engine exited early ({proc.returncode}): {output}")
        if ipc_path.exists():
            return proc
        time.sleep(0.05)
    proc.kill()
    raise RuntimeError(f"engine did not create socket {ipc_path}")


def _stop_engine(proc: subprocess.Popen[bytes], ipc_path: Path) -> None:
    proc.terminate()
    try:
        proc.wait(timeout=2)
    except subprocess.TimeoutExpired:
        proc.kill()
    if ipc_path.exists():
        ipc_path.unlink()


def _vcan0_up() -> bool:
    flags = Path("/sys/class/net/vcan0/flags")
    type_path = Path("/sys/class/net/vcan0/type")
    if not flags.is_file() or not type_path.is_file():
        return False
    try:
        if int(type_path.read_text().strip(), 0) != 280:
            return False
        return bool(int(flags.read_text().strip(), 0) & 0x1)
    except ValueError:
        return False


def test_should_flush_bounds() -> None:
    assert should_flush(0, 1.0) is False
    assert should_flush(1, 0.0) is False
    assert should_flush(1, RX_BATCH_INTERVAL_S) is True
    assert should_flush(1, RX_BATCH_MAX_INTERVAL_S) is True
    assert should_flush(RX_BATCH_MAX_FRAMES, 0.0) is True
    assert should_flush(RX_BATCH_MAX_FRAMES - 1, RX_BATCH_INTERVAL_S - 0.001) is False
    assert RX_BATCH_INTERVAL_S <= 0.016 + 1e-9
    assert RX_BATCH_MAX_INTERVAL_S <= 0.033 + 1e-9
    print("should_flush 16–33 ms / 500: ok")


def test_rx_queue_cap_and_drop_counter() -> None:
    bus = ScriptedBus()
    pump = RxPump(bus, "bus-flood", "vcan0", queue_max=RX_QUEUE_MAX)
    pump.start()
    try:
        flood = 12_000
        for i in range(flood):
            bus.inject(FakeMsg(0x100 + (i % 2048), bytes([i & 0xFF])))
        deadline = time.time() + 3.0
        while pump.dropped < flood - RX_QUEUE_MAX and time.time() < deadline:
            time.sleep(0.01)
        assert pump.qsize <= RX_QUEUE_MAX
        assert pump.queue_max == RX_QUEUE_MAX
        assert pump.dropped >= flood - RX_QUEUE_MAX
        first = pump.drain(RX_BATCH_MAX_FRAMES)
        assert len(first) <= RX_BATCH_MAX_FRAMES
        leftover = pump.drain(RX_QUEUE_MAX + 10)
        assert pump.qsize == 0
        assert len(first) + len(leftover) <= RX_QUEUE_MAX
        batch = rx_batch_message(first, pump.dropped)
        assert batch["type"] == "rx.batch"
        assert batch["payload"]["dropped"] == pump.dropped
        assert batch["payload"]["dropped"] > 0
        print(
            f"RX queue cap: qsize≤{RX_QUEUE_MAX} dropped={pump.dropped} "
            f"batch={len(first)} (visible on rx.batch)"
        )
    finally:
        pump.stop()


def test_flood_batch_timing_and_count() -> None:
    bus = ScriptedBus()
    net = _fake_sysfs_vcan0(Path(tempfile.mkdtemp(prefix="vanillabus-harden-sysfs-")))
    opened_bus = bus

    def opener(_name: str, _bitrate: int | None) -> ScriptedBus:
        return opened_bus

    manager = BusManager(sysfs_net=net, opener=opener)
    opened = manager.open("vcan0")
    try:
        inject_n = 2_400
        for i in range(inject_n):
            bus.inject(FakeMsg(0x200 + (i % 32), bytes([i & 0xFF])))
        started = time.monotonic()
        got = 0
        batches = 0
        first_flush_s: float | None = None
        deadline = time.monotonic() + 1.0
        while got < inject_n and time.monotonic() < deadline:
            frames, dropped = manager.drain_rx(RX_BATCH_MAX_FRAMES)
            if frames:
                assert len(frames) <= RX_BATCH_MAX_FRAMES
                if first_flush_s is None:
                    first_flush_s = time.monotonic() - started
                    assert first_flush_s <= RX_BATCH_MAX_INTERVAL_S + 0.05
                assert should_flush(len(frames), max(first_flush_s or 0.0, RX_BATCH_INTERVAL_S))
                got += len(frames)
                batches += 1
                assert isinstance(dropped, int)
            else:
                time.sleep(0.002)
        assert got == inject_n, f"expected {inject_n} frames, got {got}"
        assert batches >= inject_n / RX_BATCH_MAX_FRAMES
        print(
            f"flood batch: {got} frames in {batches} chunks ≤{RX_BATCH_MAX_FRAMES}, "
            f"first flush {first_flush_s * 1000:.1f} ms"
        )
    finally:
        manager.close(opened["busId"])


def test_rate_tracker_key_cap() -> None:
    cap = 64
    tracker = RateTracker(key_max=cap)
    for i in range(cap + 40):
        result = tracker.observe("bus-a", i, False, 1_000_000 + i * 1000)
        if i < cap:
            assert result is None
    assert len(tracker.stored_keys()) == cap
    # Full-size default cap stays documented.
    assert RATE_KEY_MAX == 8192
    big = RateTracker()
    for i in range(200):
        big.observe("bus-a", i, False, i)
    assert len(big.stored_keys()) == 200
    print(f"rate key cap: held {cap} after +40 unique IDs: ok")


def test_dbc_path_traversal() -> None:
    try:
        resolve_dbc_path("fixtures/dbc/../../../etc/hostname")
        raise AssertionError("expected path_not_allowed for ../ escape")
    except DbcError as exc:
        assert exc.code in {"path_not_allowed", "dbc_not_found", "dbc_invalid"}

    try:
        resolve_dbc_path("/tmp/../etc/hostname")
        raise AssertionError("expected path_not_allowed for absolute traversal")
    except DbcError as exc:
        assert exc.code in {"path_not_allowed", "dbc_not_found", "dbc_invalid"}

    link = ROOT / "fixtures" / "dbc" / f"harden-escape-{os.getpid()}.dbc"
    target = Path("/etc/hostname")
    if target.is_file() and not link.exists():
        try:
            link.symlink_to(target)
            try:
                resolve_dbc_path(str(link))
                raise AssertionError("expected path_not_allowed for symlink escape")
            except DbcError as exc:
                assert exc.code == "path_not_allowed"
            print("DBC symlink escape rejected: ok")
        finally:
            if link.exists() or link.is_symlink():
                link.unlink()
    else:
        print("DBC .. traversal rejected: ok")


def test_pump_restart_under_flood() -> None:
    bus = ScriptedBus()
    pump = RxPump(bus, "bus-restart", "vcan0", queue_max=128)
    pump.start()
    try:
        for i in range(400):
            bus.inject(FakeMsg(0x10, bytes([i & 0xFF])))
        deadline = time.time() + 1.5
        while pump.dropped == 0 and time.time() < deadline:
            time.sleep(0.01)
        assert pump.qsize <= 128
        assert pump.dropped > 0
    finally:
        pump.stop()

    pump2 = RxPump(bus, "bus-restart", "vcan0", queue_max=128)
    pump2.start()
    try:
        for i in range(40):
            bus.inject(FakeMsg(0x11, bytes([i & 0xFF])))
        deadline = time.time() + 1.0
        drained: list[dict] = []
        while len(drained) < 40 and time.time() < deadline:
            drained.extend(pump2.drain(500))
            if len(drained) < 40:
                time.sleep(0.01)
        assert len(drained) == 40
        assert pump2.qsize <= 128
        assert all(frame["can_id"] == 0x11 for frame in drained)
        print(f"pump restart under flood: resumed {len(drained)} frames, qsize≤128")
    finally:
        pump2.stop()


def test_ipc_oversized_malformed_partial() -> None:
    ipc_path = _choose_socket_path()
    proc = _start_engine(ipc_path)
    try:
        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        sock.connect(str(ipc_path))
        hello = _recv_until(sock, {"engine.hello"})
        assert hello.get("type") == "engine.hello"

        # Oversized length: engine.error then close; process stays up.
        sock.sendall(struct.pack(">I", MAX_FRAME_BYTES + 1))
        closed_or_error = False
        try:
            error = _recv_until(sock, {"engine.error"}, timeout=2.0)
            assert error.get("payload", {}).get("code") == "payload_too_large"
            closed_or_error = True
            print(f"oversized length -> {error.get('payload')}")
        except (TimeoutError, RuntimeError, ProtocolError, OSError):
            closed_or_error = True
            print("oversized length: connection closed (engine still running)")
        sock.close()
        assert closed_or_error
        if proc.poll() is not None:
            raise RuntimeError("engine crashed after oversized length")

        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        sock.connect(str(ipc_path))
        _recv_until(sock, {"engine.hello"})
        payload = b"{not-json"
        sock.sendall(struct.pack(">I", len(payload)) + payload)
        try:
            error = _recv_until(sock, {"engine.error"}, timeout=2.0)
            assert error.get("payload", {}).get("code") in {"invalid_json", "invalid_message"}
            print(f"malformed JSON -> {error.get('payload')}")
        except (TimeoutError, RuntimeError, ProtocolError, OSError):
            print("malformed JSON: connection closed (engine still running)")
        sock.close()
        if proc.poll() is not None:
            raise RuntimeError("engine crashed after malformed JSON")

        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        sock.connect(str(ipc_path))
        _recv_until(sock, {"engine.hello"})
        sock.sendall(b"\x00\x00")  # 2-byte partial header
        started = time.monotonic()
        try:
            error = _recv_until(
                sock,
                {"engine.error"},
                timeout=IPC_PARTIAL_READ_TIMEOUT_S + 1.5,
            )
            assert error.get("payload", {}).get("code") == "invalid_length"
            print(f"partial header -> {error.get('payload')}")
        except (TimeoutError, RuntimeError, ProtocolError, OSError):
            print("partial header: connection closed without hang")
        elapsed = time.monotonic() - started
        assert elapsed < IPC_PARTIAL_READ_TIMEOUT_S + 2.0, f"partial read hung ({elapsed:.1f}s)"
        sock.close()
        if proc.poll() is not None:
            raise RuntimeError("engine crashed after partial header")

        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        sock.connect(str(ipc_path))
        hello2 = _recv_until(sock, {"engine.hello"})
        assert hello2.get("type") == "engine.hello"
        sock.close()
        print("engine stayed up after oversized / malformed / partial: ok")
    finally:
        _stop_engine(proc, ipc_path)


def test_live_vcan_flood_optional() -> None:
    if not _vcan0_up():
        print(
            "SKIP vcan0 harden: vcan0 is not UP on this host. "
            "Bring it up with: sudo ./scripts/setup-vcan.sh"
        )
        return

    ipc_path = _choose_socket_path()
    proc = _start_engine(ipc_path)
    try:
        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        sock.connect(str(ipc_path))
        _recv_until(sock, {"engine.hello"})
        _send_message(sock, {"type": "bus.open", "id": "open-h", "payload": {"name": "vcan0"}})
        opened = _recv_until(sock, {"bus.open", "engine.error"})
        if opened.get("type") != "bus.open":
            raise AssertionError(f"bus.open failed: {opened!r}")

        try:
            import can
        except ImportError:
            can = None
        if can is None:
            print("SKIP vcan0 harden flood: python-can not importable")
            sock.close()
            return

        tx = can.Bus(interface="socketcan", channel="vcan0")
        try:
            for i in range(80):
                tx.send(
                    can.Message(
                        arbitration_id=0x42B,
                        data=bytes([i & 0xFF, 0x22, 0x33, 0x44]),
                        is_extended_id=False,
                    )
                )
        finally:
            tx.shutdown()

        started = time.monotonic()
        seen = 0
        deadline = time.time() + 1.0
        while time.time() < deadline and seen < 1:
            remaining = deadline - time.time()
            try:
                message = _recv_until(sock, {"rx.batch"}, timeout=max(remaining, 0.01))
            except TimeoutError:
                break
            frames = message.get("payload", {}).get("frames")
            assert isinstance(frames, list)
            assert len(frames) <= RX_BATCH_MAX_FRAMES
            assert isinstance(message.get("payload", {}).get("dropped"), int)
            seen += sum(1 for frame in frames if frame.get("can_id") == 0x42B)
        assert seen >= 1, "expected live rx.batch during vcan flood"
        assert time.monotonic() - started <= 0.5
        print(f"live vcan flood: {seen} frames, batches ≤{RX_BATCH_MAX_FRAMES}: ok")
        sock.close()
    finally:
        _stop_engine(proc, ipc_path)


def main() -> int:
    test_should_flush_bounds()
    test_rx_queue_cap_and_drop_counter()
    test_flood_batch_timing_and_count()
    test_rate_tracker_key_cap()
    test_dbc_path_traversal()
    test_pump_restart_under_flood()
    test_ipc_oversized_malformed_partial()
    test_live_vcan_flood_optional()
    print("PASS")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
